library/lmots.c

```diff
@@ -0,0 +1,855 @@
+static int public_key_from_hashed_digit_array( const mbedtls_lmots_parameters_t *params,
+                                               const unsigned char *y_hashed_digits,
+                                               unsigned char *pub_key )
+{
+    psa_hash_operation_t op;
+    psa_status_t status;
+    size_t output_hash_len;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+
+    op = psa_hash_operation_init( );
+    status = psa_hash_setup( &op, PSA_ALG_SHA_256 );
+    ret = mbedtls_lms_error_from_psa( status );
+    if( ret != 0 )
+        goto exit;
+
+    status = psa_hash_update( &op,
+                              params->I_key_identifier,
+                              MBEDTLS_LMOTS_I_KEY_ID_LEN );
+    ret = mbedtls_lms_error_from_psa( status );
+    if( ret != 0 )
+        goto exit;
+
+    status = psa_hash_update( &op, params->q_leaf_identifier,
+                              MBEDTLS_LMOTS_Q_LEAF_ID_LEN );
+    ret = mbedtls_lms_error_from_psa( status );
+    if( ret != 0 )
+        goto exit;
+
+    status = psa_hash_update( &op, D_PUBLIC_CONSTANT_BYTES, D_CONST_LEN );
+    ret = mbedtls_lms_error_from_psa( status );
+    if( ret != 0 )
+        goto exit;
+
+    status = psa_hash_update( &op, y_hashed_digits,
+                              MBEDTLS_LMOTS_P_SIG_DIGIT_COUNT(params->type) *
+                              MBEDTLS_LMOTS_N_HASH_LEN(params->type) );
+    ret = mbedtls_lms_error_from_psa( status );
+    if( ret != 0 )
+        goto exit;
+
+    status = psa_hash_finish( &op, pub_key,
+                              MBEDTLS_LMOTS_N_HASH_LEN(params->type),
+                              &output_hash_len );
+    ret = mbedtls_lms_error_from_psa( status );
+
+exit:
+    psa_hash_abort( &op );
+    return( ret );
+}
+
+int mbedtls_lms_error_from_psa( psa_status_t status )
+{
+    switch( status )
+    {
+        case PSA_SUCCESS:
+            return( 0 );
+        case PSA_ERROR_HARDWARE_FAILURE:
+            return( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+        case PSA_ERROR_NOT_SUPPORTED:
+            return( MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED );
+        case PSA_ERROR_BUFFER_TOO_SMALL:
+            return( MBEDTLS_ERR_LMS_BUFFER_TOO_SMALL );
+        case PSA_ERROR_INVALID_ARGUMENT:
+            return( MBEDTLS_ERR_LMS_BAD_INPUT_DATA );
+        default:
+            return( MBEDTLS_ERR_ERROR_GENERIC_ERROR );
+    }
+}
+
+void mbedtls_lmots_init_public( mbedtls_lmots_public_t *ctx )
+{
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_lmots_public_t  ) ) ;
+}
+
+void mbedtls_lmots_free_public( mbedtls_lmots_public_t *ctx )
+{
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_lmots_public_t  ) ) ;
+}
+
+int mbedtls_lmots_import_public_key( mbedtls_lmots_public_t *ctx,
+                                 const unsigned char *key, size_t key_len )
+{
+    ctx->params.type =
+        mbedtls_lms_network_bytes_to_unsigned_int( MBEDTLS_LMOTS_TYPE_LEN,
+                key + MBEDTLS_LMOTS_SIG_TYPE_OFFSET );
```

Comment:
This feels like another case (along with the other case with the OFFSET pointer arithmetic) where an always-on runtime check feels a little unnecessary. We've got the check that the key buffer isn't smaller than a LMOTS public key, so what we really care about here is that `LMOTS_SIG_TYPE_OFFSET` isn't greater than the size of an LMOTS public key. Potentially, we can `static_assert` this for all LMS types (since we only have one) but it feel weird to only check these two offset values, since this function overwriting is no more dangerous than, say, the memcpy two lines below. Thoughts?

As a side-note, this should definitely be `PUBLIC_KEY_TYPE_OFFSET` :)

Reply:
This is the first instruction in a non-static function. Nothing has been checked yet. `key_len` is a variable so `static_assert` is not relevant.

*After* the check `MBEDTLS_LMOTS_PUBLIC_KEY_LEN(ctx->params.type)`, `key + x` is valid for any `x <= MBEDTLS_LMOTS_PUBLIC_KEY_LEN(ctx->params.type)`. And it's ok if we know that a specific value of `x` is less than the bound only because the `MBEDTLS_LMOTS_xxx` constants have been defined appropriately. But not before this check.

Reply:
Ah yes I think I misunderstood where this comment applied to, I'll add a check.